# Getting more movie features using TMDB API with a randomly selected sample dataset (N = 50) from `movie_feb2025.csv`

## Roadmap: What this notebook does  

**Goal:** build a clean, feature-rich dataset that can be fed to an LLM / ML model to **predict movie ratings** (initial focus: TMDB `vote_average`).  

Below is the step-by-step flow you’ll see in this notebook:

| Step | Section | What happens & why |
|------|---------------------------|--------------------|
| **1** | **Load & Inspect Raw TMDB Data** | Load the original CSV → explore column meanings → quantify missing values. This tells us which features already exist and their data quality. |
| **2** | **Split Rated vs Unrated → Train / Test** | Keep only movies with a non-zero TMDB rating. Split them 80 / 20 into `train` and `test`. From the **train** set randomly sample **N = 20** rows that we’ll enrich end-to-end (quick dev loop). |
| **3** | **Enrich Movie-level Features** | Via **TMDB movie** endpoints pull extra signal:  `Credits`  · `Keywords`  · `Reviews`  · `Watch Providers`  · `External IDs`. |
| **4** | **Enrich Person-level Features** | Use every `actor_*_id` & `crew_*_id` to hit **TMDB people** endpoints. Collect the most predictive fields (chiefly `biography`, plus `gender`, `popularity`, etc.). |
| **5** | **Enrich External Features** | **IMDb:** scrape `imdb_rating` / `imdb_votes` / `imdb_metascore`.  **Wikidata:** scrape awards & nominations → `award_count`, `award_names`, `nominated_count`, `nominated_names`. |
| **6** | **Feature Cleaning & Transformation** | Convert raw objects to model-friendly text or numeric forms (e.g. genre arrays → “Action, Drama”; language codes → “English”). Drop columns we no longer need. |
| **7** | **Column Re-ordering** | Rearrange all features into a logical, human-readable order (IDs & meta first, numerics next, person blocks grouped by actor_1, actor_2, …). |
| **8** | **Save Final Dataset** | Export the fully enriched, cleaned CSV that’s ready for finetuning or downstream ML. |

## Load and Inspect the Dataset

In [1]:
import pandas as pd
df = pd.read_csv('movie_feb2025.csv', on_bad_lines='skip', engine='python')
df['popularity'] = pd.to_numeric(df['popularity'], errors='coerce')

print(f"Successfully loaded. Total number of rows: {len(df)}")
print(f"Actual number of columns: {df.shape[1]}")
print(df.head(5))

Successfully loaded. Total number of rows: 1032235
Actual number of columns: 27
   adult                     backdrop_path  \
0  False  /hQ4pYsIbP22TMXOUdSfC2mjWrO0.jpg   
1  False  /l94l89eMmFKh7na2a1u5q67VgNx.jpg   
2  False  /f2t4JbUvQIjUF5FstG1zZFAp02N.jpg   
3  False  /iUUpKunmBN5l8goObADBaFHnxQ8.jpg   
4  False                               NaN   

                               belongs_to_collection    budget  \
0  {'id': 1382526, 'name': "Kaurismäki's Proletar...         0   
1  {'id': 1382526, 'name': "Kaurismäki's Proletar...         0   
2                                                NaN   4000000   
3                                                NaN  21000000   
4                                                NaN     42000   

                                              genres  \
0  [{'id': 35, 'name': 'Comedy'}, {'id': 18, 'nam...   
1  [{'id': 35, 'name': 'Comedy'}, {'id': 18, 'nam...   
2                     [{'id': 35, 'name': 'Comedy'}]   
3  [{'id': 28, 'name':

In [2]:
df.columns

Index(['adult', 'backdrop_path', 'belongs_to_collection', 'budget', 'genres',
       'homepage', 'id', 'imdb_id', 'origin_country', 'original_language',
       'original_title', 'overview', 'popularity', 'poster_path',
       'production_companies', 'production_countries', 'release_date',
       'revenue', 'runtime', 'spoken_languages', 'status', 'tagline', 'title',
       'video', 'vote_average', 'vote_count', 'novelty'],
      dtype='object')

In [3]:
df.dtypes

adult                     object
backdrop_path             object
belongs_to_collection     object
budget                    object
genres                    object
homepage                  object
id                       float64
imdb_id                   object
origin_country            object
original_language         object
original_title            object
overview                  object
popularity               float64
poster_path               object
production_companies      object
production_countries      object
release_date              object
revenue                  float64
runtime                  float64
spoken_languages          object
status                    object
tagline                   object
title                     object
video                     object
vote_average             float64
vote_count               float64
novelty                  float64
dtype: object

In [4]:
df.isna().sum()

adult                          0
backdrop_path             705978
belongs_to_collection    1002926
budget                      5956
genres                      5956
homepage                  905222
id                          5956
imdb_id                   432462
origin_country              5956
original_language           5956
original_title              7062
overview                  178927
popularity                  8466
poster_path               263570
production_companies        7211
production_countries        8466
release_date              112419
revenue                     8466
runtime                     8466
spoken_languages            8466
status                      8466
tagline                   879456
title                       8475
video                       8466
vote_average                8466
vote_count                  8466
novelty                  1032235
dtype: int64

In [5]:
missing_info = pd.DataFrame({
    'missing_count': df.isna().sum(),
    'missing_percent': (df.isna().sum() / len(df) * 100).round(2)
}).sort_values(by='missing_percent', ascending=False)

print(missing_info)

                       missing_count  missing_percent
novelty                      1032235           100.00
belongs_to_collection        1002926            97.16
homepage                      905222            87.70
tagline                       879456            85.20
backdrop_path                 705978            68.39
imdb_id                       432462            41.90
poster_path                   263570            25.53
overview                      178927            17.33
release_date                  112419            10.89
revenue                         8466             0.82
title                           8475             0.82
vote_count                      8466             0.82
popularity                      8466             0.82
vote_average                    8466             0.82
production_countries            8466             0.82
video                           8466             0.82
runtime                         8466             0.82
spoken_languages            

### Drop Entries with Missing TMDB IDs

In [6]:
df = df[~df['id'].isna()]

In [7]:
actural_na_id = (df['id'] == 0).sum()/len(df) * 100
print(f"{actural_na_id:.2f}% of movies have an TMDB id of NA")

0.12% of movies have an TMDB id of NA


In [8]:
df = df[df['id'] != 0]

In [9]:
df.isna().sum()

adult                          0
backdrop_path             700022
belongs_to_collection     996491
budget                         0
genres                         0
homepage                  898939
id                             0
imdb_id                   426506
origin_country                 0
original_language              0
original_title                11
overview                  172971
popularity                  1265
poster_path               257614
production_companies        1255
production_countries        1265
release_date              105218
revenue                     1265
runtime                     1265
spoken_languages            1265
status                      1265
tagline                   872255
title                       1274
video                       1265
vote_average                1265
vote_count                  1265
novelty                  1025034
dtype: int64

In [10]:
percentage_zero = (df['vote_average'] == 0).sum()/len(df) * 100
print(f"{percentage_zero:.2f}% of movies have a vote_average of 0")

63.19% of movies have a vote_average of 0


### Evaluate Feature Usefulness for Prediction

In [11]:
missing_info = pd.DataFrame({
    'missing_count': df.isna().sum(),
    'missing_percent': (df.isna().sum() / len(df) * 100).round(2)
}).sort_values(by='missing_percent', ascending=False)

print(missing_info)

                       missing_count  missing_percent
novelty                      1025034           100.00
belongs_to_collection         996491            97.22
homepage                      898939            87.70
tagline                       872255            85.10
backdrop_path                 700022            68.29
imdb_id                       426506            41.61
poster_path                   257614            25.13
overview                      172971            16.87
release_date                  105218            10.26
revenue                         1265             0.12
vote_count                      1265             0.12
vote_average                    1265             0.12
popularity                      1265             0.12
video                           1265             0.12
production_companies            1255             0.12
production_countries            1265             0.12
title                           1274             0.12
runtime                     

In [12]:
df['budget'] = pd.to_numeric(df['budget'], errors='coerce')
actural_na_budget = (df['budget'] == 0).sum()/len(df) * 100
print(f"{actural_na_budget:.2f}% of movies have a budget of 0")

94.03% of movies have a budget of 0


In [13]:
actural_na_genres = (df['genres'] == '[]').sum()/len(df) * 100
print(f"{actural_na_genres:.2f}% of movies have a genres of NA")

28.13% of movies have a genres of NA


In [14]:
actural_na_popularity = (df['popularity'] == '0').sum()/len(df) * 100
print(f"{actural_na_popularity:.2f}% of movies have a popularity of 0")

0.00% of movies have a popularity of 0


In [15]:
actural_na_production_companies = (df['production_companies'] == '[]').sum()/len(df) * 100
print(f"{actural_na_production_companies:.2f}% of movies have a production_companies of NA")

53.07% of movies have a production_companies of NA


In [16]:
actural_na_production_countries = (df['production_countries'] == '[]').sum()/len(df) * 100
print(f"{actural_na_production_countries:.2f}% of movies have a production_countries of NA")

38.56% of movies have a production_countries of NA


In [17]:
actural_na_revenue = (df['revenue'] == 0).sum()/len(df) * 100
print(f"{actural_na_revenue:.2f}% of movies have a revenue of 0")

97.59% of movies have a revenue of 0


In [18]:
actural_na_runtime = (df['runtime'] == 0).sum()/len(df) * 100
print(f"{actural_na_runtime:.2f}% of movies have a runtime of 0")

23.23% of movies have a runtime of 0


In [19]:
actural_na_spoken_languages = (df['spoken_languages'] == '[]').sum()/len(df) * 100
print(f"{actural_na_spoken_languages:.2f}% of movies have a spoken_languages of NA")

35.75% of movies have a spoken_languages of NA


In [20]:
actural_na_vote_average = (df['vote_average'] == 0).sum()/len(df) * 100
print(f"{actural_na_vote_average:.2f}% of movies have a vote_average of 0")

63.19% of movies have a vote_average of 0


In [21]:
actural_na_vote_count = (df['vote_count'] == 0).sum()/len(df) * 100
print(f"{actural_na_vote_count:.2f}% of movies have a vote_count of 0")

63.16% of movies have a vote_count of 0


### Overview of Current Features

Based on the [TMDB API documentation MOVIES - Details](https://developer.themoviedb.org/reference/movie-details), the following introduces the current dataset columns, their meaning, missing rate, and whether they are useful for predicting.

`adult`: 0% missing. Defaults to False. However, some movies may be incorrectly marked as non-adult due to unavailable data. Interpret with caution.

`backdrop_path`: 68.21% missing. Although visual feature are not needed for training, its presence may correlate with a movie’s rating. Can be transformed into a binary feature.

`belongs_to_collection`: 97.14% missing. Only a small percentage of movies belong to a collection, but its presence may correlate with a movie’s rating. May be useful if cleaned.

`budget`: 0% missing. Defaults to 0. 94.03% are zeros, which often indicates unavailable data rather than an actual zero budget. Interpret with caution.

`genres`: 28.19% actual missing. May be useful if cleaned.

`homepage`: 87.62% missing. While the webpage content is not used for training, the presence of a homepage may correlate with a movie’s rating. Can be transformed into a binary feature.

`id`: 0% missing after cleaning. This is the unique identifier for each movie in the TMDB database.

`imdb_id`: 41.56% missing. While the IMDB content is not used for training, the presence of it may correlate with a movie’s rating. 

`origin_country`: 0% missing. This feature may correlate with a movie’s rating.

`original_language`: 0 missing. This feature may correlate with a movie’s rating.

`original_title`: 0.11% missing. This feature may reflect cultural or regional factors that influence ratings.

`overview`: 16.85% missing. This is a key input feature for training and predicting ratings.

`popularity`: 0.24% missing. This is a dynamic metric provided by TMDB, calculated based on factors such as number of votes, views, favorites, watchlists, release date, and previous popularity scores.

Since it combines several other variables already present in the dataset (e.g. release_date), it may introduce **multicollinearity** issues if used together with them. Use with caution.

`poster_path`: 25.10% missing. Although visual feature are not needed for training, its presence may correlate with a movie’s rating. Can be transformed into a binary feature.

`production_companies`: 53.07% actual missing. May be useful if cleaned.

`production_countries`: 38.56% actual missing. May be useful if cleaned.

`release_date`: 10.37% missing. This feature may correlate with a movie’s rating.

`revenue`: 0.24% missing, but 97.59% are zeros, which often indicates unavailable data rather than an actual zero revenue. Interpret with caution.

`runtime`: 23.23% actual missing. This feature may correlate with a movie’s rating.

`spoken_languages`: 35.75% actual missing. May be useful if cleaned.

`status`: 0.12% missing. Categorical (e.g., Released). Could be relevant to ratings.

`tagline`: 85.11% missing. Optional marketing line. It may carry signal for well-marketed films.

`title`: 0.25% missing. Some of the movies have english tranlated title, some of them don't. It could have affect the potential rating of the movie.

`video`: 0.24% missing. Defaults to false. It could have affect the potential rating of the movie.

`vote_average`: 63.19% missing. Defaults to 0 when unrated. This is a target variable for prediction.

`vote_count`: 63.16% missing. Defaults to 0 when unrated. This is another target variable for prediction.

Note: Image-related features such as `backdrop_path` and `poster_path` are not full URLs. To view the actual image, prepend the path with https://image.tmdb.org/t/p/original/ according to [TMDB's image guide](https://developer.themoviedb.org/docs/image-basics).  

Example:  
https://image.tmdb.org/t/p/original/hQ4pYsIbP22TMXOUdSfC2mjWrO0.jpg

## Split the Dataset

**63.19%** of the movies in the current dataset have a `vote_average` of zero on TMDB.

Since our goal is to build a predictive model that estimates user ratings based on existing movie features, we only retain movies with valid ratings (vote_average ≠ 0) for model training and evaluation.

The dataset is therefore divided into two subsets:

- Unrated movies (vote_average == 0): These are excluded from model training but can later be used as input for prediction.

- Rated movies (vote_average ≠ 0): These form the basis of our supervised machine learning task. 

We further split the Rated movies subset as follows:
- **80%** for the training dataset
- **20%** for the test dataset

We randomly draw a small sample of N = 10 movies from the training set and use them to run the code that gathers additional movie features.

In [22]:
unrated = df[df['vote_average'] == 0]

In [23]:
rated = df[df['vote_average'] != 0]

In [24]:
from sklearn.model_selection import train_test_split
train, test = train_test_split(rated, test_size=0.2, random_state=42)

In [25]:
sample = train.sample(n=49, random_state=42)
specific_row = train[train["id"] == 238]
sample = pd.concat([sample, specific_row], ignore_index=True)

*The Godfather* (ID: 238) is intentionally included in the sample dataset, as it provides complete entries across all relevant features, such as `reviews_TMDB` and `imdb_metascore`, which are often missing in less popular titles. It helps improve the robustness of the code.

In [26]:
sample.head()

adult                     backdrop_path belongs_to_collection  budget  \
0  False  /eHSemCAEY6TyWO7r1erblh7Xgmw.jpg                   NaN  1000.0   
1  False  /tMFMMFXGOC6nQIuDSy3ihThRNgk.jpg                   NaN     0.0   
2  False                               NaN                   NaN     0.0   
3  False  /cjuwqee9OOY9btzf4OO0WiF2f63.jpg                   NaN     0.0   
4  False                               NaN                   NaN     0.0   

                                              genres homepage         id  \
0                      [{'id': 18, 'name': 'Drama'}]      NaN  1084823.0   
1                [{'id': 99, 'name': 'Documentary'}]      NaN   801172.0   
2                                                 []      NaN  1175181.0   
3  [{'id': 99, 'name': 'Documentary'}, {'id': 107...      NaN   985789.0   
4                     [{'id': 27, 'name': 'Horror'}]      NaN  1305928.0   

      imdb_id origin_country original_language  ... revenue runtime  \
0  tt14837634         ['GB']                en  ...     0.0     5.0   
1         NaN   ['FR', 'CH']                fr  ...     0.0     0.0   
2   tt1163322         ['AU']                en  ...     0.0    72.0   
3         NaN         ['US']                en  ...     0.0    78.0   
4  tt16408474         ['US']                en  ...     0.0    90.0   

                                    spoken_languages    status  \
0  [{'english_name': 'English', 'iso_639_1': 'en'...  Released   
1  [{'english_name': 'French', 'iso_639_1': 'fr',...  Released   
2                                                 []  Released   
3                                                 []  Released   
4  [{'english_name': 'English', 'iso_639_1': 'en'...  Released   

                tagline                                title  video  \
0  Is anything missing?                         A Phone Call  False   
1                   NaN  Cyberbully, When the Haters Rise Up  False   
2                   NaN                                Taxi!  False   
3                   NaN               The English Civil Wars   True   
4                   NaN                             The Cult  False   

   vote_average  vote_count novelty  
0          10.0         1.0     NaN  
1           6.8         2.0     NaN  
2           3.5         1.0     NaN  
3          10.0         1.0     NaN  
4           8.0         7.0     NaN  

[5 rows x 27 columns]

In [27]:
sample.to_csv("sample.csv", index=False)

## Enrich Movie-level Features

For each movie, the following TMDB API endpoints were used to extract additional metadata and enrich the dataset:

`Credits`: cast and crew details

`Keywords`: thematic tags and topics

`Reviews`: short user reviews with author info

`Watch Providers`: platforms where the movie can be streamed or bought

`External IDs`: IMDb, Wikidata, Facebook, Instagram, Twitter

### Credits

Based on the [TMDB API documentation for movie credits](https://developer.themoviedb.org/reference/movie-credits), the following variables were extracted and appended to the dataset:

- `cast_count`: number of cast members  
- `crew_count`: number of crew members  
- `cast`: full JSON object containing cast details  
- `crew`: full JSON object containing crew details  

Each object inside `cast` and `crew` contains structured attributes such as:

- `adult`, `gender`, `id`, `known_for_department`, `name`, `original_name`, `popularity`, `profile_path`, `credit_id`  
- (cast only): `cast_id`, `character`, `order`  
- (crew only): `department`, `job`

In [28]:
import requests
import time
import json

# Load the sample dataset containing movie IDs
df = pd.read_csv("sample.csv")

# API credentials
api_key = "7b4782c9b0a5abfc789b2b79cfab2601"
bearer_token = "eyJhbGciOiJIUzI1NiJ9.eyJhdWQiOiI3YjQ3ODJjOWIwYTVhYmZjNzg5YjJiNzljZmFiMjYwMSIsIm5iZiI6MTc0Mzc1NDE0Mi43NjQ5OTk5LCJzdWIiOiI2N2VmOTM5ZTBjNzkxYmViNTdhY2Y1YWUiLCJzY29wZXMiOlsiYXBpX3JlYWQiXSwidmVyc2lvbiI6MX0.2FmX4__p3XQQi5IagAJ1Csa-3WvL67-msP6158CbsHs"

# Prepare empty lists for new columns
cast_count_list = []
crew_count_list = []
cast_raw_list = []
crew_raw_list = []

# Loop through each movie ID in the dataset and retrieve credits data
for movie_id in df["id"]:
    url = f"https://api.themoviedb.org/3/movie/{int(movie_id)}/credits?language=en-US"
    headers = {
        "accept": "application/json",
        "Authorization": f"Bearer {bearer_token}"
    }

    try:
        response = requests.get(url, headers=headers)
        if response.status_code == 200:
            data = response.json()
            cast = data.get("cast", [])
            crew = data.get("crew", [])

            cast_count_list.append(len(cast))
            crew_count_list.append(len(crew))
            cast_raw_list.append(json.dumps(cast))  # Store raw JSON string
            crew_raw_list.append(json.dumps(crew))
        else:
            # Handle API error gracefully
            cast_count_list.append(None)
            crew_count_list.append(None)
            cast_raw_list.append(None)
            crew_raw_list.append(None)
    except Exception:
        # Handle other unexpected errors
        cast_count_list.append(None)
        crew_count_list.append(None)
        cast_raw_list.append(None)
        crew_raw_list.append(None)

    time.sleep(0.25)  # Respect API rate limits

# Add new columns to the original DataFrame
df["cast_count"] = cast_count_list
df["crew_count"] = crew_count_list
df["cast"] = cast_raw_list
df["crew"] = crew_raw_list

# Save the updated CSV file
df.to_csv("sample.csv", index=False)

### Keywords

Based on the [TMDB API documentation for movie keywords](https://developer.themoviedb.org/reference/movie-keywords), the `keywords` field was added.  

It contains thematic tags such as `espionage`, `based on true story`, etc., useful for training.

In [29]:
# Initialize a list to store keywords for each movie
keywords_data = []

# Loop through each movie ID
for movie_id in df["id"]:
    url = f"https://api.themoviedb.org/3/movie/{int(movie_id)}/keywords"
    headers = {
        "accept": "application/json",
        "Authorization": f"Bearer {bearer_token}"
    }

    try:
        response = requests.get(url, headers=headers)
        if response.status_code == 200:
            data = response.json()
            keywords = [k["name"] for k in data.get("keywords", [])]
            keywords_data.append(keywords)
        else:
            keywords_data.append(None)
    except Exception as e:
        keywords_data.append(None)

    time.sleep(0.25)  # Avoid rate limiting

# Add keywords as a new column
df["keywords"] = keywords_data

# Save the updated CSV file
df.to_csv("sample.csv", index=False)

### Reviews

Based on the [TMDB API documentation for movie reviews](https://developer.themoviedb.org/reference/movie-reviews), the `reviews` field was added.

Although the TMDB database provides relatively few user reviews, we choose to retain the available review data, as it may carry signals that correlate with a movie’s rating. This feature is included with appropriate handling of missing values.

In [30]:
# Initialize a list to store review lists per movie
review_data = []

# Loop through each movie ID
for movie_id in df["id"]:
    all_reviews = []
    page = 1

    while True:
        url = f"https://api.themoviedb.org/3/movie/{int(movie_id)}/reviews?language=en-US&page={page}"
        headers = {
            "accept": "application/json",
            "Authorization": f"Bearer {bearer_token}"
        }

        try:
            response = requests.get(url, headers=headers)
            if response.status_code == 200:
                data = response.json()
                results = data.get("results", [])
                
                if not results:
                    break

                for item in results:
                    author = item.get("author", "unknown")
                    content = item.get("content", "").replace("\r", " ").replace("\n", " ").strip()
                    all_reviews.append(f"{author}: {content}")
                
                if page >= data.get("total_pages", 1):
                    break
                else:
                    page += 1
                    time.sleep(0.3)  # 控制速率
            else:
                all_reviews = None
                break
        except Exception as e:
            all_reviews = None
            break

    review_data.append(all_reviews if all_reviews else None)
    time.sleep(0.3)

df["reviews_TMDB"] = review_data

df.to_csv("sample.csv", index=False)

### Watch Providers

Based on the [TMDB API documentation for watch providers](https://developer.themoviedb.org/reference/movie-watch-providers), the `watch_providers` field was added.  

It contains information about where a movie can be streamed, rented, or purchased across different countries and platforms (e.g., Netflix, Apple TV, Amazon). May be weakly correlated with rating due to platform selection bias and audience exposure.

In [31]:
# Initialize a list to store watch provider info
watch_provider_data = []

# Loop through movie IDs
for movie_id in df["id"]:
    url = f"https://api.themoviedb.org/3/movie/{int(movie_id)}/watch/providers"
    headers = {
        "accept": "application/json",
        "Authorization": f"Bearer {bearer_token}"
    }

    try:
        response = requests.get(url, headers=headers)
        if response.status_code == 200:
            data = response.json()
            # Store full country-wise results for now
            watch_provider_data.append(data.get("results", {}))
        else:
            watch_provider_data.append(None)
    except Exception as e:
        watch_provider_data.append(None)

    time.sleep(0.25)  # Prevent hitting the API rate limit

# Add the data to the dataframe
df["watch_providers"] = watch_provider_data

# Save the updated CSV file
df.to_csv("sample.csv", index=False)

### External IDs

Based on the [TMDB API documentation for movie external IDs](https://developer.themoviedb.org/reference/movie-external-ids), the `external_ids` field was appended to the dataset. 

It includes external reference identifiers such as IMDb ID, Wikidata ID, and social media handles (Facebook, Instagram, Twitter).

* **IMDb ID** will later be used to scrape additional features including IMDb Rating, IMDb Vote Count, and Metascore.

* **Wikidata ID** will be used to scrape awards and nominations through structured queries.

* Social media accounts (e.g., number of followers, engagement, posting activity) may serve as indirect indicators of marketing investment. However, due to the high cost of data collection, restricted access, and their relatively low expected contribution to rating prediction, we chose not to include them at this stage.

In [32]:
# List to hold external IDs for each movie
external_ids_data = []

# Loop through each movie ID
for movie_id in df["id"]:
    url = f"https://api.themoviedb.org/3/movie/{int(movie_id)}/external_ids"
    headers = {
        "accept": "application/json",
        "Authorization": f"Bearer {bearer_token}"
    }

    try:
        response = requests.get(url, headers=headers)
        if response.status_code == 200:
            data = response.json()
            external_ids_data.append(data)  # Store full dict (includes IMDb, Facebook, Twitter, etc.)
        else:
            external_ids_data.append(None)
    except Exception:
        external_ids_data.append(None)

    time.sleep(0.25)  # Avoid hitting rate limit

# Add as a new column to the dataframe
df["external_ids"] = external_ids_data

# Save the updated CSV file
df.to_csv("sample.csv", index=False)

## Enrich Person-level Features

### Extract and Structure Top Cast Information

Only the top 5 actors are included, as they usually represent the core cast with the greatest influence on audience perception and overall film quality. 

In [33]:
# Load the original dataset
sample = pd.read_csv("sample.csv")

# Define the number of top cast members to retain
MAX_CAST = 5

# Fields to extract from each actor (excluding 'order')
BASE_FIELDS = ["id", "name", "gender", "popularity"]

# Create new columns in the DataFrame for each actor and field
for idx in range(1, MAX_CAST + 1):
    for field in BASE_FIELDS:
        sample[f"actor_{idx}_{field}"] = None

# Loop through each row to extract top 5 cast members' data
for row_idx, cast_json in sample["cast"].items():
    try:
        cast_list = json.loads(cast_json)
    except (TypeError, json.JSONDecodeError):
        cast_list = []

    # Sort cast by 'order' and keep the top N
    top_cast = sorted(cast_list, key=lambda x: x.get("order", 999))[:MAX_CAST]

    # Populate the new columns with selected fields
    for pos, actor in enumerate(top_cast, start=1):
        for field in BASE_FIELDS:
            sample.at[row_idx, f"actor_{pos}_{field}"] = actor.get(field)

# Save the updated CSV file
sample.to_csv("sample.csv", index=False)
print("Expanded cast columns have been saved to sample.csv")

Expanded cast columns have been saved to sample.csv


### Extract and Structure Top Crew Information

To prioritize the most influential crew members for predicting a movie’s rating, we focus on six key roles that are most commonly associated with major film awards, roles that most significantly influence a film’s quality and rating.

In [34]:
# Load the original dataset
sample = pd.read_csv("sample.csv")

# Award-driven priority crew jobs and limits per job
priority_jobs = {
    "Director": 1,                 # Best Director
    "Writer": 1,                   # Best Original / Adapted Screenplay
    "Producer": 1,                 # Best Picture (producer credit)
    "Director of Photography": 1,  # Best Cinematography
    "Editor": 1,                   # Best Film Editing
    "Original Music Composer": 1   # Best Original Score
}

BASE_FIELDS = ["id", "name", "gender", "popularity"]

def prefix(job: str) -> str:
    """Convert job title to a safe lowercase prefix for column names."""
    return job.lower().replace(" ", "_").replace("/", "_").replace("-", "_")

# Pre-create empty columns for each (job, idx, field)
for job, max_keep in priority_jobs.items():
    p = prefix(job)
    for idx in range(1, max_keep + 1):
        for field in BASE_FIELDS:
            sample[f"{p}_{idx}_{field}"] = None

# Parse each crew JSON list, keep only priority jobs, fill columns
for row_idx, crew_json in sample["crew"].items():
    try:
        crew_list = json.loads(crew_json)
    except (TypeError, json.JSONDecodeError):
        crew_list = []

    # Group crew by job
    grouped = {}
    for member in crew_list:
        job_name = member.get("job")
        if job_name in priority_jobs:
            grouped.setdefault(job_name, []).append(member)

    # Fill DataFrame up to N per job
    for job_name, max_keep in priority_jobs.items():
        members = grouped.get(job_name, [])[:max_keep]
        p = prefix(job_name)
        for idx, person in enumerate(members, start=1):
            for field in BASE_FIELDS:
                sample.at[row_idx, f"{p}_{idx}_{field}"] = person.get(field)

# Save the updated CSV file
sample.to_csv("sample.csv", index=False)

### Use cast id to get more features

In [35]:
# Load the original dataset
df = pd.read_csv("sample.csv")

bearer_token = (
    "eyJhbGciOiJIUzI1NiJ9.eyJhdWQiOiI3YjQ3ODJjOWIwYTVhYmZjNzg5YjJiNzljZmFiMjYwMSIsIm5iZiI6MTc0Mzc1NDE0Mi43NjQ5OTk5LCJzdWIiOiI2N2VmOTM5ZTBjNzkxYmViNTdhY2Y1YWUiLCJzY29wZXMiOlsiYXBpX3JlYWQiXSwidmVyc2lvbiI6MX0.2FmX4__p3XQQi5IagAJ1Csa-3WvL67-msP6158CbsHs"
)

BASE_URL  = "https://api.themoviedb.org/3/person/{}?language=en-US"
HEADERS   = {"accept": "application/json", "Authorization": f"Bearer {bearer_token}"}
SLEEP_SEC = 0.25

# Cache to avoid duplicate network calls

detail_cache = {}

def fetch_person(pid: int) -> dict:
    """Return TMDB person-details JSON (empty dict if any error)."""
    if pd.isna(pid):
        return {}
    if pid in detail_cache:
        return detail_cache[pid]

    try:
        r = requests.get(BASE_URL.format(int(pid)), headers=HEADERS, timeout=10)
        detail_cache[pid] = r.json() if r.status_code == 200 else {}
    except Exception:
        detail_cache[pid] = {}

    time.sleep(SLEEP_SEC)
    return detail_cache[pid]

actor_id_cols = [c for c in df.columns if c.startswith("actor_") and c.endswith("_id")]

# Expand actor_N_details columns (N = 1…5)
for id_col in actor_id_cols:
    details_col = id_col.replace("_id", "_details")
    df[details_col] = df[id_col].apply(
        lambda x: json.dumps(fetch_person(x), ensure_ascii=False)
    )

# Save the updated CSV file
df.to_csv("sample.csv", index=False)

#### Extract biography column for actors

In [36]:
df = pd.read_csv("sample.csv")

detail_cols = [
    c for c in df.columns
    if c.startswith("actor_") and c.endswith("_details")
]

def extract_bio(detail_json):
    """Return biography string or None."""
    if pd.isna(detail_json) or detail_json == "":
        return None
    try:
        return json.loads(detail_json).get("biography")
    except (json.JSONDecodeError, TypeError):
        return None

for col in detail_cols:
    bio_col = col.replace("_details", "_biography")
    df[bio_col] = df[col].apply(extract_bio)

df.to_csv("sample.csv", index=False)

### Use crew id to get more features

In [37]:
df = pd.read_csv("sample.csv")

bearer_token = (
    "eyJhbGciOiJIUzI1NiJ9.eyJhdWQiOiI3YjQ3ODJjOWIwYTVhYmZjNzg5YjJiNzljZmFiMjYwMSIsIm5iZiI6MTc0Mzc1NDE0Mi43NjQ5OTk5LCJzdWIiOiI2N2VmOTM5ZTBjNzkxYmViNTdhY2Y1YWUiLCJzY29wZXMiOlsiYXBpX3JlYWQiXSwidmVyc2lvbiI6MX0.2FmX4__p3XQQi5IagAJ1Csa-3WvL67-msP6158CbsHs"
)
HEADERS = {"accept": "application/json",
           "Authorization": f"Bearer {bearer_token}"}
BASE_URL = "https://api.themoviedb.org/3/person/{}?language=en-US"
SLEEP = 0.25            # ≤ 4 req/s (safe for TMDB)

# PERSON-DETAILS FETCHER (with simple cache)
cache = {}

def get_person(pid):
    if pd.isna(pid):
        return {}
    pid = int(pid)
    if pid in cache:
        return cache[pid]
    try:
        r = requests.get(BASE_URL.format(pid), headers=HEADERS, timeout=10)
        cache[pid] = r.json() if r.status_code == 200 else {}
    except Exception:
        cache[pid] = {}
    time.sleep(SLEEP)
    return cache[pid]

# IDENTIFY ALL CREW-ID COLUMNS
crew_id_cols = []
for job, n in priority_jobs.items():
    p = prefix(job)                       # e.g. director_, writer_
    for i in range(1, n + 1):
        col = f"{p}_{i}_id"
        if col in df.columns:             # cast columns are actor_*, ignored
            crew_id_cols.append(col)

# CREATE *_details COLUMNS
for id_col in crew_id_cols:
    details_col = id_col.replace("_id", "_details")
    # only fetch where details col is missing to save API quota
    df[details_col] = df[id_col].apply(
        lambda x: json.dumps(get_person(x), ensure_ascii=False)
    )

df.to_csv("sample.csv", index=False)

#### Extract biography collumn for crews

In [38]:
df = pd.read_csv("sample.csv")

def extract_bio(js):
    if pd.isna(js) or js == "": 
        return None
    try:
        return json.loads(js).get("biography")
    except (json.JSONDecodeError, TypeError):
        return None

crew_detail_cols = [c for c in df.columns 
                    if c.endswith("_details") and not c.startswith("actor_")]

for col in crew_detail_cols:
    df[col.replace("_details", "_biography")] = df[col].apply(extract_bio)

df.to_csv("sample.csv", index=False)

## Enrich Enrich External Features

### Clean the External ID Column

In this step, we retain only the IMDb and Wikidata IDs, as they provide access to relevant information such as ratings and awards. Social media IDs (Facebook, Instagram, Twitter) are excluded because their added value for rating prediction is likely limited.

In [39]:
import ast

df = pd.read_csv("sample.csv")

def to_dict(x):
    if pd.isna(x) or x == "":
        return {}
    if isinstance(x, dict):
        return x
    try:
        return ast.literal_eval(x)
    except Exception:
        return {}

df["imdb_id"] = df["external_ids"].apply(lambda x: to_dict(x).get("imdb_id"))
df["wikidata_id"] = df["external_ids"].apply(lambda x: to_dict(x).get("wikidata_id"))

df.to_csv("sample.csv", index=False)

### Extract IMDB Features

In this step, we extract three key features from IMDb:

* `imdb_rating` captures the general public’s evaluation of the film, providing an external benchmark to TMDB ratings.

* `imdb_votes` reflects the popularity of a movie among viewers, which often correlates with cultural impact or marketing reach.

* `imdb_metascore` introduces a critical perspective, as it aggregates reviews from professional critics. This feature can help balance user-driven ratings with expert evaluations.

These three metrics together offer a more comprehensive view of a movie’s reception, which is valuable for improving the accuracy of our rating prediction model.

In [40]:
import pandas as pd, requests, re, json, time
from bs4 import BeautifulSoup

df = pd.read_csv("sample.csv") 

UA = {"User-Agent":
      "Mozilla/5.0 (Windows NT 10.0; Win64; x64) "
      "AppleWebKit/537.36 (KHTML, like Gecko) "
      "Chrome/125.0 Safari/537.36",
      "Accept-Language": "en-US,en;q=0.5"}
BASE = "https://www.imdb.com/title/{}/"
SLEEP = 0.5
cache = {}

def fetch_html(imdb_id):
    """Download page; if redirected to 'consent' once, follow the link."""
    r = requests.get(BASE.format(imdb_id), headers=UA, timeout=10)
    # If IMDb shows a EU consent page (title 'consentstv'), follow the continue link
    if "consent" in r.url and "continue" in r.text.lower():
        soup = BeautifulSoup(r.text, "html.parser")
        cont = soup.find("a", href=True)
        if cont and cont["href"].startswith("/title/"):
            r = requests.get("https://www.imdb.com" + cont["href"], headers=UA, timeout=10)
    return r.text

def scrape(imdb_id):
    if pd.isna(imdb_id):          # empty cell
        return None, None, None
    if imdb_id in cache:
        return cache[imdb_id]

    rating = votes = metascore = None
    try:
        html = fetch_html(imdb_id)
        soup = BeautifulSoup(html, "html.parser")

        # rating & votes from JSON-LD
        ld_tag = soup.find("script", type="application/ld+json")
        if ld_tag:
            try:
                data = json.loads(ld_tag.string)
                agg = data.get("aggregateRating", {})
                rating = float(agg.get("ratingValue")) if agg.get("ratingValue") else None
                votes  = int(agg.get("ratingCount"))  if agg.get("ratingCount") else None
            except (ValueError, TypeError):
                pass

        # metascore
        score = soup.select_one("span.score-meta")
        if score and score.text.strip().isdigit():
            metascore = int(score.text.strip())
        else:
            alt = soup.select_one('[data-testid="metacritic-score-box"]')
            if alt and re.search(r"\d+", alt.text):
                metascore = int(re.search(r"\d+", alt.text).group())
            else:
                m = re.search(r'"metacritic"\s*:\s*{[^}]*?"score"\s*:\s*(\d+)', html)
                if m:
                    metascore = int(m.group(1))

    except Exception:
        pass

    cache[imdb_id] = (rating, votes, metascore)
    time.sleep(SLEEP)
    return rating, votes, metascore

df[["imdb_rating", "imdb_votes", "imdb_metascore"]] = (
    df["imdb_id"]
      .apply(lambda x: pd.Series(scrape(x),
                                 index=["imdb_rating", "imdb_votes", "imdb_metascore"]))
)

df.to_csv("sample.csv", index=False)

### Extract Wikidata features

While Wikidata contains many other attributes (e.g., release date, director, language), these have already been captured in the TMDB dataset.

In this step, we focus solely on two key features from Wikidata: the number and names of awards received (`award_count`, `award_names`) and nominations (`nominated_count`, `nominated_names`).

In [41]:
"""
Add four columns to sample.csv, based on the Wikidata ID in column
`wikidata_id`.

new columns:
    award_count          – distinct awards won  (P166)
    award_names          – “; ”-joined English labels of those awards
    nominated_count      – distinct award nominations (P1411)
    nominated_names      – “; ”-joined English labels of those nominations
"""

df = pd.read_csv("sample.csv")

API = "https://www.wikidata.org/wiki/Special:EntityData/{}.json"
LABEL_API = "https://www.wikidata.org/w/api.php"
SLEEP = 0.4
item_cache, label_cache = {}, {}

def get_entity(qid):
    if qid in item_cache:                       # cached JSON
        return item_cache[qid]
    try:
        data = requests.get(API.format(qid), timeout=10).json()
        item_cache[qid] = data
        time.sleep(SLEEP)
        return data
    except Exception:
        item_cache[qid] = {}
        return {}

def get_labels(qids):
    """Return dict {qid: en_label}, fetching uncached items in bulk."""
    need = [q for q in qids if q not in label_cache]
    if need:
        params = {
            "action": "wbgetentities",
            "ids": "|".join(need),
            "format": "json",
            "props": "labels",
            "languages": "en"
        }
        try:
            obj = requests.get(LABEL_API, params=params, timeout=10).json()
            for q, ent in obj.get("entities", {}).items():
                label_cache[q] = ent.get("labels", {}).get("en", {}).get("value")
            time.sleep(SLEEP)
        except Exception:
            for q in need:
                label_cache[q] = None
    return {q: label_cache.get(q) for q in qids}

def extract_awards(qid):
    data = get_entity(qid)
    claims = data.get("entities", {}).get(qid, {}).get("claims", {})
    won = {snak["mainsnak"]["datavalue"]["value"]["id"]
           for snak in claims.get("P166", [])        # award received
           if snak.get("mainsnak", {}).get("datavalue")}
    nom = {snak["mainsnak"]["datavalue"]["value"]["id"]
           for snak in claims.get("P1411", [])       # nominated for
           if snak.get("mainsnak", {}).get("datavalue")}
    labels = get_labels(won | nom)
    won_names = sorted(filter(None, (labels[q] for q in won)))
    nom_names = sorted(filter(None, (labels[q] for q in nom)))
    return (
        len(won_names), "; ".join(won_names) if won_names else None,
        len(nom_names), "; ".join(nom_names) if nom_names else None
    )

df[["award_count", "award_names",
    "nominated_count", "nominated_names"]] = (
    df["wikidata_id"]
      .apply(lambda x: pd.Series(extract_awards(str(x)) if pd.notna(x) else (None, None, None, None),
                                 index=["award_count", "award_names",
                                        "nominated_count", "nominated_names"]))
)

df.to_csv("sample.csv", index=False)

## Feature Cleaning & Transformation

The above steps have acquired all the potentially useful features in across TMDB and external platforms, so the next step will be make the output sheet cleaner for futher fine-tuning process.

### Transform to Binary Features

#### backdrop_path & poster_path

In [42]:
df = pd.read_csv("sample.csv")

# backdrop_path → has_backdrop
df["has_backdrop"] = df["backdrop_path"].notna() & df["backdrop_path"].str.strip().ne("")

# poster_path → has_poster
df["has_poster"] = df["poster_path"].notna() & df["poster_path"].str.strip().ne("")

df.to_csv("sample.csv", index=False)

#### homepage

In [43]:
df = pd.read_csv("sample.csv")

df["has_homepage"] = df["homepage"].apply(lambda x: True if pd.notna(x) and x.strip() != "" else False)

df.to_csv("sample.csv", index=False)

### Basic Data Cleaning

#### belongs_to_collection

In [44]:
df = pd.read_csv("sample.csv")

def extract_collection_name(val):
    if pd.isna(val) or val.strip() == "":
        return None
    try:
        parsed = ast.literal_eval(val)
        return parsed.get("name")
    except (ValueError, SyntaxError):
        return None

df["collection_name"] = df["belongs_to_collection"].apply(extract_collection_name)

df.to_csv("sample.csv", index=False)

#### genres

In [45]:
df = pd.read_csv("sample.csv")

def extract_genre_names(val):
    if pd.isna(val) or val.strip() == "":
        return None
    try:
        genre_list = ast.literal_eval(val)
        if isinstance(genre_list, list):
            names = [g["name"] for g in genre_list if "name" in g]
            return ", ".join(names)
        return None
    except (ValueError, SyntaxError):
        return None

df["genres_names"] = df["genres"].apply(extract_genre_names)

df.to_csv("sample.csv", index=False)

#### origin_country

In [46]:
!pip install pycountry
import pycountry

df = pd.read_csv("sample.csv")

def code_to_country(code):
    try:
        country = pycountry.countries.get(alpha_2=code)
        return country.name if country else code
    except:
        return code

def parse_and_convert(val):
    if pd.isna(val) or val.strip() == "":
        return None
    try:
        country_codes = ast.literal_eval(val)
        if not isinstance(country_codes, list):
            return None
        countries = [code_to_country(c) for c in country_codes]
        return ", ".join(countries)
    except (ValueError, SyntaxError):
        return None

df["origin_countries"] = df["origin_country"].apply(parse_and_convert)
df.to_csv("sample.csv", index=False)

#### original_language

In [47]:
df = pd.read_csv("sample.csv")

def code_to_language(code):
    try:
        lang = pycountry.languages.get(alpha_2=code)
        return lang.name if lang else code
    except:
        return code

df["original_language_name"] = df["original_language"].apply(code_to_language)

df.to_csv("sample.csv", index=False)

#### production_countries

In [48]:
df = pd.read_csv("sample.csv")

def extract_country_names(val):
    if pd.isna(val):
        return None
    try:
        items = ast.literal_eval(val)
        if isinstance(items, list):
            names = [item.get("name") for item in items if isinstance(item, dict) and item.get("name")]
            return ", ".join(names) if names else None
    except:
        return None
    return None

df["production_country_names"] = df["production_countries"].apply(extract_country_names)

df.to_csv("sample.csv", index=False)

#### spoken_languages

In [49]:
df = pd.read_csv("sample.csv")

def extract_language_names(val):
    if pd.isna(val):
        return None
    try:
        items = ast.literal_eval(val)
        if isinstance(items, list):
            names = [item.get("english_name") for item in items if isinstance(item, dict) and item.get("english_name")]
            return ", ".join(names) if names else None
    except:
        return None
    return None

df["spoken_language_names"] = df["spoken_languages"].apply(extract_language_names)

df.to_csv("sample.csv", index=False)

#### production_companies

In [50]:
df = pd.read_csv("sample.csv")

def extract_company_names(val):
    if pd.isna(val):
        return None
    try:
        items = ast.literal_eval(val)
        if isinstance(items, list):
            names = [item.get("name") for item in items if isinstance(item, dict) and item.get("name")]
            return ", ".join(names) if names else None
    except:
        return None
    return None

df["production_company_names"] = df["production_companies"].apply(extract_company_names)

df.to_csv("sample.csv", index=False)

#### keywords

In [51]:
df = pd.read_csv("sample.csv")

def clean_keywords(val):
    if pd.isna(val):
        return None
    try:
        keywords = ast.literal_eval(val)
        if isinstance(keywords, list):
            return ", ".join(str(k) for k in keywords)
    except:
        return None
    return None

df["keyword_list"] = df["keywords"].apply(clean_keywords)

df.to_csv("sample.csv", index=False)

#### watch_providers

In [52]:
df = pd.read_csv("sample.csv")

def extract_provider_names(val):
    if pd.isna(val) or val.strip() == "":
        return None
    try:
        data = ast.literal_eval(val)            # dict keyed by country code
        if not isinstance(data, dict):
            return None
        names = []
        for country in data.values():           # loop each country's sub-dict
            if not isinstance(country, dict):
                continue
            for cat in ("flatrate", "rent", "buy"):
                for provider in country.get(cat, []):
                    n = provider.get("provider_name")
                    if n:
                        names.append(n)
        if not names:
            return None
        return ", ".join(sorted(set(names)))
    except Exception:
        return None

df["watch_provider_names"] = df["watch_providers"].apply(extract_provider_names)

df.to_csv("sample.csv", index=False)

#### gender-related columns

In [53]:
df = pd.read_csv("sample.csv")

# mapping: numeric code → descriptive label
gender_map = {
    0: "Not set / not specified",
    1: "Female",
    2: "Male",
    3: "Non-binary"
}

# find every column whose name ends with "_gender"
gender_cols = [c for c in df.columns if c.endswith("_gender")]

for col in gender_cols:
    label_col = col.replace("_gender", "_gender_label")
    df[label_col] = df[col].map(gender_map)

df.to_csv("sample.csv", index=False)

#### reviews_TMDB

In [54]:
df = pd.read_csv("sample.csv")

def clean_reviews(val):
    if pd.isna(val) or val.strip() == "":
        return None
    try:
        reviews = ast.literal_eval(val)          # "['user: text', ...]" → list
        texts = []
        for r in reviews:
            if not isinstance(r, str):
                continue
            body = r.split(":", 1)[-1].strip()
            if body:
                texts.append(body)
        return " ".join(texts) if texts else None
    except Exception:
        return None

df["reviews_clean"] = df["reviews_TMDB"].apply(clean_reviews)

df.to_csv("sample.csv", index=False)

## Column Re-ordering

### Remove Uncleaned and Untransformed Columns

In [55]:
import pandas as pd

df = pd.read_csv("sample.csv")

columns_to_drop = [
    "backdrop_path", "belongs_to_collection", "genres", "homepage", "origin_country",
    "original_language", "poster_path", "production_companies", "production_countries",
    "spoken_languages", "cast", "crew", "keywords", "watch_providers", "external_ids",
    "actor_1_details", "actor_2_details", "actor_3_details", "actor_4_details", "actor_5_details",
    "director_1_details", "writer_1_details", "producer_1_details",
    "director_of_photography_1_details", "editor_1_details", "original_music_composer_1_details",
    "actor_1_gender", "actor_2_gender", "actor_3_gender", "actor_4_gender", "actor_5_gender",
    "director_1_gender", "writer_1_gender", "producer_1_gender",
    "director_of_photography_1_gender", "editor_1_gender", "original_music_composer_1_gender"
]

df.drop(columns=[col for col in columns_to_drop if col in df.columns], inplace=True)

df.to_csv("sample.csv", index=False)

### Reorder and Save

In [56]:
df = pd.read_csv("sample.csv")

identifiers = [
    "id", "imdb_id", "wikidata_id"
]

basic_info = [
    "title", "original_title", "tagline", "overview",
    "release_date", "status",
]

numeric_info = [
    "budget", "revenue", "runtime",
    "popularity", "vote_average", "vote_count",
    "imdb_rating", "imdb_votes", "imdb_metascore", "novelty", 
]

flags = [
    "adult", "video", "has_homepage", 
    "has_backdrop", "has_poster", 
]

language_geo = [
    "original_language_name", "origin_countries",
    "production_country_names", "spoken_language_names",
]

categorical_text = [
    "collection_name", "genres_names",
    "keyword_list", "watch_provider_names",
    "production_company_names",
]

award_cols = [
    "award_count", "award_names",
    "nominated_count", "nominated_names",
]

count_cols = [
    "cast_count", "crew_count",
]

review_cols = [
    "reviews_TMDB",
]

people_prefixes = [
    "actor",
    "director",
    "writer",
    "producer",
    "director_of_photography",
    "editor",
    "original_music_composer",
]

def gather_people_cols(prefix):
    pat = re.compile(rf"^{re.escape(prefix)}_(\d+)_")
    cols = [c for c in df.columns if c.startswith(prefix + "_")]
    ordered = []
    idx_set = sorted({int(pat.match(c).group(1)) for c in cols if pat.match(c)})
    for idx in idx_set:
        ordered.extend([c for c in cols if pat.match(c) and int(pat.match(c).group(1)) == idx])
    return ordered

people_cols_ordered = []
for pre in people_prefixes:
    people_cols_ordered.extend(gather_people_cols(pre))

ordered = (
    identifiers
    + basic_info
    + numeric_info
    + flags
    + language_geo
    + categorical_text
    + award_cols
    + count_cols
    + review_cols
    + people_cols_ordered
)

remaining = [c for c in df.columns if c not in ordered]
ordered += remaining

df = df[ordered]
df.to_csv("sample_ordered.csv", index=False)